City-of-Helsinki/helsinki-paatokset
# UHF-8858: import org chart

# [UHF-8858](https://helsinkisolutionoffice.atlassian.net/browse/UHF-8858)


## What was done


* Import organization data to Drupal, add service for querying organization hierarchy.

## How to install

* Make sure your instance is up and running on correct branch.
  * `git checkout UHF-8858-import-org-chart`
  * `make new`
* Run `make drush-cr`
* Import policymakers: `drush migrate-import ahjo_decisionmakers:all --update`.

## How to test


* [x] Importing all organizations (takes a while):
      ```
      drush migrate-import ahjo_organizations -v
      ```
      or run with `--limit` option.
      ```
      drush migrate-import ahjo_organizations -v --limit 100
      ```
* [x] Re-import policymakers:
      ```
      drush migrate-import ahjo_decisionmakers:all --update
      ```
      This should create links between policymaker and organization nodes.
* [x] View changes to [`/public/modules/custom/paatokset_ahjo_proxy/src/Commands/AhjoAggregatorCommands.php`](https://github.com/City-of-Helsinki/helsinki-paatokset/blob/UHF-8858-import-org-chart/public/modules/custom/paatokset_ahjo_proxy/src/Commands/AhjoAggregatorCommands.php).
      Try running the drush command:
      ```
      drush ahjo-proxy:fix-decisionmaker-references -v
      ```
      I think the command should not do anything unless you have the production database.
      
* [x] Check out the admin organization list at [`/admin/content/organizations`](https://helsinki-paatokset.docker.so/en/admin/content/organizations?langcode=All).
* [x] View changes to [`/docker/openshift/crons/run-aggregated-ahjo-integrations.sh`](https://github.com/City-of-Helsinki/helsinki-paatokset/blob/UHF-8858-import-org-chart/docker/openshift/crons/run-aggregated-ahjo-integrations.sh)
* [x] Try fetching the organization hierarchy for some decisionmakers. Start php console:
  ```
  drush php:cli
  ```
  Try running the following code:
  ```php
  use Drupal\paatokset_policymakers\Service\OrganizationService;
  use Drupal\paatokset_policymakers\Service\PolicymakerService;

  $nodeStorage = \Drupal::entityTypeManager()->getStorage('node');
  $orgService = \Drupal::service(OrganizationService::class);

  $council = $nodeStorage->loadByProperties([
    'langcode' => 'fi',
    'type' => 'policymaker',
    'field_policymaker_id' => PolicymakerService::CITY_COUNCIL_DM_ID 
  ]);
  $council = reset($council);

  $org = $orgService->getPolicymakerOrganization($council, $council->language());
  $hierarchy = $orgService->getOrganizationHierarchy($org);
  ```

  


[UHF-8858]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-8858?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
